cpp/serve/engine_actions/action_commons.cc

```diff
@@ -66,6 +66,14 @@ void ActionStepPostProcess(Array<Request> requests, EngineState estate, Array<Mo
       continue;
     }
 
+    // Update the grammar matcher state if it exists.
+    if (rstate->mstates[0]->grammar_state_matcher) {
+      const auto& grammar_state_matcher = rstate->mstates[0]->grammar_state_matcher.value();
+      for (auto token_id : delta_token_ids) {
+        grammar_state_matcher->AcceptToken(token_id);
+      }
+    }
+
```

Comment:
I am curious about the reason we update the matcher here. Is it because of the stop string handling? I think the state update logic is more appropriate to be put in methods like `CommitTokens`, `AddDraftTokens`, etc. (see below), which dedicate to update states (including adding new tokens and rollback tokens). This function here mainly wants to figure out the delta generated token for callback streaming, and process the finished requests.

https://github.com/mlc-ai/mlc-llm/blob/f00b70e1739591597e617f22a0a7c6352deee715/cpp/serve/request_state.cc#L49-L75

Reply:
I suppose right now JSON generation does not support speculative inference (or speculative inference does not support JSON generation), since we don't use `Rollback` of the grammar state matcher for ReqeustState. We may want to think about how to bring this in the future.

Reply:
Yes speculative inference is not supported now. That is a feature we will consider in the next round of updates.